### ISSUE TITLE
methylInheritance

### ISSUE THREAD

**Issue by adeschen** [Bioconductor/Contributions#265]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/adeschen/methylInheritance

Confirm the following by editing each check box to '[x]'

- [X] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [X] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [X] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [X] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [X] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [X] The 'devel' branch for new packages and features.
- [X] The stable 'release' branch, made available every six
      months, for bug fixes.
- [X] _Bioconductor_ version control using [Subversion][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/source-control/
[6]: http://bioconductor.org/developers/how-to/git-mirrors/

**Comment by hpages** [comment]
Hi Astrid,

Thanks for this new contribution to Bioconductor. Overall the package looks good but unfortunately the man pages are cluttered with repeated information and typos that hurt its usability. See details below. Please address at your earliest convenience.

Thanks!
H.

1. Please merge the man pages for `runPermutationUsingMethylKitInfo()` and `runPermutationUsingRDSFile()` in a single man page. Right now the 2 big man pages are almost identical and repeating so much information in 2 different places is bad practise. It's called the DRY principle (Don't Repeat Yourself) in software engineering. This principle applies to the code and the documentation.

2. Same thing for the man pages for `runObservationUsingMethylKitInfo()` and `runObservationUsingRDSFile()`.

3. Furthermore, there is no need to have these `*MethylKitInfo()`/`*RDSFile()` combos. Please replace the 4 functions with 2 functions e.g. `runPermutation()` (or `runPermutationAnalysis()`) for the 1st combo and `runObservation()` (or `runObservationAnalysis()`) for the 2nd combo.

4. Please provide *running* examples in the man pages for `runPermutation()` and `runObservation()` i.e. they should not be put inside `\dontrun` directives and they should work properly. For example right now if I try to run the example for `runPermutationUsingMethylKitInfo()` I get the following:
```
    > permutationResults <- runPermutationUsingMethylKitInfo(
          samplesForTransgenerationalAnalysis,
          type="sites", nbrPermutations=3, vSeed=221))
    Warning messages:
    1: glm.fit: fitted probabilities numerically 0 or 1 occurred 
    2: glm.fit: algorithm did not converge 
    3: glm.fit: fitted probabilities numerically 0 or 1 occurred 
    4: glm.fit: fitted probabilities numerically 0 or 1 occurred 
    5: glm.fit: fitted probabilities numerically 0 or 1 occurred 
    Warning messages:
    1: glm.fit: fitted probabilities numerically 0 or 1 occurred 
    2: glm.fit: fitted probabilities numerically 0 or 1 occurred 
    3: glm.fit: fitted probabilities numerically 0 or 1 occurred 
    4: glm.fit: fitted probabilities numerically 0 or 1 occurred 
    5: glm.fit: fitted probabilities numerically 0 or 1 occurred 
    6: glm.fit: fitted probabilities numerically 0 or 1 occurred 
    7: glm.fit: fitted probabilities numerically 0 or 1 occurred 
    Warning messages:
    1: glm.fit: fitted probabilities numerically 0 or 1 occurred 
    2: glm.fit: algorithm did not converge 
    3: glm.fit: fitted probabilities numerically 0 or 1 occurred 
```
  I'm not sure if all those warnings are expected but they are definitely surprising/confusing and will make most users question the correctness of the software.

5. Applying the DRY principle again, you should not repeat the description of the internal structure of a methylInheritanceAllResults object in every man page documenting a function that returns such object. This description should be done only once (e.g. in the man page for `mergePermutationAndObservation()`) and the man pages of other functions that return such object (e.g. `loadAllRDSResults()`, `runPermutationUsingMethylKitInfo()`, `runPermutationUsingRDSFile()`) should refer to that description.

6. Your man pages contain many typos that make them hard to read/understand. For example:
    - a \code{list} of class \code{} when ...
    - @return a \code{list} containing...
    - itemize (instead of \itemize
    etc...
Please re-read carefully all your man pages and fix these problems (or have someone read them and report the problems to you).

7. Please indent your code properly so it's easier to read/understand e.g.
```
     runPermutationUsingMethylKitInfo(
         samplesForTransgenerationalAnalysis,
         type="sites", nbrPermutations=3, vSeed=221)
```
  instead of
```
     runPermutationUsingMethylKitInfo(methylKitInfo =
     samplesForTransgenerationalAnalysis, type = "sites",
     nbrPermutations = 3, vSeed = 221)
```

8. Misspelled  functions names in the vignette: `runPermutationUsingMehtylKitInfo()` and `runObservationUsingMehtylKitInfo()` (these functions don't exist).

9. Doing
```
    dir(system.file("extdata", package = "methylInheritance"),
        pattern = "resultsForTransgenerationalAnalysis.RDS",
        full.names = TRUE)
```
is an obfuscated way of doing
```
    system.file("extdata", "resultsForTransgenerationalAnalysis.RDS",
                package="methylInheritance")
```
Please keep your code as simple and straightforward as possible.

10. Direct slot access is considered bad practice and is strongly discouraged e.g. showing this
```
    demoForTransgenerationalAnalysis[[2]]@.Data[[1]]
```
  in the vignette is not needed or useful so please remove.
  For this one (also from the vignette):
```
    slot(demoForTransgenerationalAnalysis[[1]], "treatment")
```
  An accessor function should be used instead. More precisely the methylKit package should provide a `treatment()` getter to let the user access the "treatment" slot of a methylRawList object. I realize that the methylKit package is maintained by other people but I encourage you to contact them to ask them to add the `treatment()` getter to their package. In the mean time please use
```
    demoForTransgenerationalAnalysis[[1]]@treatment
```
  instead of the above.

11. You have many man pages that document things that are not exposed (i.e. not exported) to the user. For example: createDataStructure.Rd, createOutputDir.Rd, getGRangesFromMethylDiff.Rd, interGeneration.Rd, etc... Is this intended?

**Comment by adeschen** [comment]
Hi Hervé,

We have corrected all the listed issues. See the corrections/explanations below.

Thanks,
Astrid

**1.** Please merge the man pages for `runPermutationUsingMethylKitInfo()` and `runPermutationUsingRDSFile()` in a single man page. Right now the 2 big man pages are almost identical and repeating so much information in 2 different places is bad practise. It's called the DRY principle (Don't Repeat Yourself) in software engineering. This principle applies to the code and the documentation.

**Correction:** The merging of the `runPermutationUsingMethylKitInfo()` and `runPermutationUsingRDSFile()` functions into one function has solved this issue.

**2.** Same thing for the man pages for `runObservationUsingMethylKitInfo()` and `runObservationUsingRDSFile()`.

**Correction**: The merging of the `runObservationUsingMethylKitInfo()` and `runObservationUsingRDSFile()` functions into one function has solved this issue.

**3.** Furthermore, there is no need to have these `*MethylKitInfo()/*RDSFile()` combos. Please replace the 4 functions with 2 functions e.g. `runPermutation()` (or `runPermutationAnalysis()`) for the 1st combo and `runObservation()` (or `runObservationAnalysis()`) for the 2nd combo.

**Correction**: The `runPermutationUsingMethylKitInfo()` and `runPermutationUsingRDSFile()` functions have been merged into one function called `runPermutation()`. The `runObservationUsingMethylKitInfo()` and `runObservationUsingRDSFile()` functions have been merged into one function called `runObservation()`.

**4.** Please provide running examples in the man pages for `runPermutation()` and `runObservation()` i.e. they should not be put inside `\dontrun` directives and they should work properly. For example right now if I try to run the example for `runPermutationUsingMethylKitInfo()` I get the following:

```
permutationResults <- runPermutationUsingMethylKitInfo(
          samplesForTransgenerationalAnalysis,
          type="sites", nbrPermutations=3, vSeed=221))
    Warning messages:
    1: glm.fit: fitted probabilities numerically 0 or 1 occurred 
    2: glm.fit: algorithm did not converge 
    3: glm.fit: fitted probabilities numerically 0 or 1 occurred 
    4: glm.fit: fitted probabilities numerically 0 or 1 occurred 
    5: glm.fit: fitted probabilities numerically 0 or 1 occurred 
    Warning messages:
    1: glm.fit: fitted probabilities numerically 0 or 1 occurred 
    2: glm.fit: fitted probabilities numerically 0 or 1 occurred 
    3: glm.fit: fitted probabilities numerically 0 or 1 occurred 
    4: glm.fit: fitted probabilities numerically 0 or 1 occurred 
    5: glm.fit: fitted probabilities numerically 0 or 1 occurred 
    6: glm.fit: fitted probabilities numerically 0 or 1 occurred 
    7: glm.fit: fitted probabilities numerically 0 or 1 occurred 
    Warning messages:
    1: glm.fit: fitted probabilities numerically 0 or 1 occurred 
    2: glm.fit: algorithm did not converge 
    3: glm.fit: fitted probabilities numerically 0 or 1 occurred 
```


**Correction**: The warning messages are generated by the `calculateDiffMeth()` function from Bioconductor `methylKit` package. Since the example section of the `calculateDiffMeth()` man page is also generating those warning messages, we thought it was appropriate to have the same approach. The
warning messages are not related to a problematic situation. Those messages are generated when 0 and 1 values are perfectly split between cases and controls in a logistic regression. Since methylation levels are often expected to be 0% (0) or 100% (1) and genome-wide testing implies a lot of regression tests, the situation of perfect 0/1 split is likely to happen. To remove the warning messages, the `calculateDiffMeth()` function has been embedded inside the `suppressWarnings()` function.

**5.** Applying the DRY principle again, you should not repeat the description of the internal structure of a `methylInheritanceAllResults` object in every man page documenting a function that returns such object. This description should be done only once (e.g. in the man page for `mergePermutationAndObservation()`) and the man pages of other functions that return such object (e.g. `loadAllRDSResults()`, `runPermutationUsingMethylKitInfo()`, `runPermutationUsingRDSFile()`) should refer to that description.

**Correction**: The main description of the returned object is now in the `mergePermutationAndObservation()` man page. The man pages of the related functions contain a short description and a @seealso link that refers to the `mergePermutationAndObservation()` man page.

**6.** Your man pages contain many typos that make them hard to read/understand. For example:
- a \code{list} of class \code{} when ...
- @return a \code{list} containing...
- itemize (instead of \itemize
- etc...
Please re-read carefully all your man pages and fix these problems (or have someone read them and report the problems to you).

**Correction**: Typos have been corrected. Vignette and man pages have been reviewed.

**7.** Please indent your code properly so it's easier to read/understand e.g.
     
```
runPermutationUsingMethylKitInfo(
        samplesForTransgenerationalAnalysis,
        type="sites", nbrPermutations=3, vSeed=221)
```
instead of

```
runPermutationUsingMethylKitInfo(methylKitInfo =
samplesForTransgenerationalAnalysis, type = "sites",
nbrPermutations = 3, vSeed = 221)
```


**Correction**: Indentation has been corrected.

**8.** Misspelled functions names in the vignette: `runPermutationUsingMehtylKitInfo()` and `runObservationUsingMehtylKitInfo()` (these functions don't exist).

**Correction**: Misspelled functions names have been corrected.

**9.**  Doing

`    dir(system.file("extdata", package = "methylInheritance"),
        pattern = "resultsForTransgenerationalAnalysis.RDS",
        full.names = TRUE)`

is an obfuscated way of doing

`    system.file("extdata", "resultsForTransgenerationalAnalysis.RDS",
                package="methylInheritance")`

Please keep your code as simple and straightforward as possible.

**Correction:** Code  has been modified and is not using `system.file()` function.

**10.** Direct slot access is considered bad practice and is strongly discouraged e.g. showing this

    `demoForTransgenerationalAnalysis[[2]]@.Data[[1]]`

in the vignette is not needed or useful so please remove.
For this one (also from the vignette):

    `slot(demoForTransgenerationalAnalysis[[1]], "treatment")`

An accessor function should be used instead. More precisely the `methylKit` package should provide a `treatment()` getter to let the user access the "treatment" slot of a `methylRawList` object. I realize that the `methylKit` package is maintained by other people but I encourage you to contact them to ask them to add the `treatment()` getter to their package. In the mean time please use

    `demoForTransgenerationalAnalysis[[1]]@treatment`

instead of the above.

**Correction:** The `getTreatment()` function from `methylKit` package is now replacing the direct slot access. 

**11.** You have many man pages that document things that are not exposed (i.e. not exported) to the user. For example: createDataStructure.Rd, createOutputDir.Rd, getGRangesFromMethylDiff.Rd, interGeneration.Rd, etc... Is this intended?

**Explanation**: Those man pages are related to internal functions. Those are not indented to be seen by users but are useful for maintenance.

**Comment by hpages** [comment]
Hi Astrid,
Thanks for addressing these issues. I'm marking the package as accepted. Congratulations!
In the next few days, we will add the package to our Subversion repository and to our daily builds. We will send you important information about this.
Regards,
H.
